src/aiy/assistant/device_helpers.py

```diff
@@ -0,0 +1,117 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Device registration helpers for the Google Assistant API."""
+
+import json
+import os
+import random
+
+import google.auth.transport
+
+import aiy.assistant.auth_helpers
+
+
+_DEVICE_MODEL = "voice-kit"
+_DEVICE_MANUFACTURER = "AIY Projects"
+_DEVICE_NAME = "Voice Kit"
+_DEVICE_TYPE = "action.devices.types.LIGHT"
+
+_DEVICE_ID_FILE = os.path.join(
+        aiy.assistant.auth_helpers._VR_CACHE_DIR, 'device_id.json')
+
+
+def _get_project_id():
+    with open(aiy.assistant.auth_helpers._ASSISTANT_CREDENTIALS_FILE) as f:
+        client_secrets_data = json.load(f)
+        return client_secrets_data["installed"]["project_id"]
+
+
+def _get_api_url(*args):
+    return "/".join(
+            ("https://embeddedassistant.googleapis.com/v1alpha2/projects",) + args)
+
+def _load_ids(id_path):
+    with open(id_path, 'r') as f:
+        id_data = json.load(f)
+    return id_data["model_id"], id_data["device_id"]
+
+
+def _save_ids(id_path, model_id, device_id):
+    if not os.path.exists(os.path.dirname(id_path)):
+        os.makedirs(os.path.dirname(id_path))
+
+    id_data = {
+            "model_id": model_id,
+            "device_id": device_id,
+    }
+    with open(id_path, 'w') as f:
+        json.dump(id_data, f)
+
+
+def _get_model_id(credentials, session, project_id):
+    model_id = "%s-%s" % (project_id, _DEVICE_MODEL)
+    payload = {
+            "device_model_id": model_id,
+            "project_id": project_id,
+            "device_type": _DEVICE_TYPE,
+            "manifest": {
+                    "manufacturer": _DEVICE_MANUFACTURER,
+                    "product_name": _DEVICE_NAME,
+            },
+    }
+    r = session.post(_get_api_url(project_id, "deviceModels"),
+                     data=json.dumps(payload))
+    if r.status_code == 409:
+        # Already exists: update
+        r = session.put(_get_api_url(project_id, "deviceModels", model_id),
+                        data=json.dumps(payload))
+    r.raise_for_status()
+    return model_id
+
+
+def get_ids(credentials, model_id=None):
+    """get_ids gets a Device ID for use with the Google Assistant SDK.
+
+    It optionally also gets a Device Model ID if one is not given. The IDs are
+    cached on disk so that a device keeps a consistent ID.
+
+    Returns:
+        a tuple: (device_id, model_id)
+    """
+
+    if os.path.exists(_DEVICE_ID_FILE):
+        return _load_ids(_DEVICE_ID_FILE)
+
+    session = google.auth.transport.requests.AuthorizedSession(credentials)
+    project_id = _get_project_id()
+    model_id = model_id or _get_model_id(credentials, session, project_id)
+
+    device_id = "%s-%06d" % (model_id, random.randrange(1000000))
+    payload = {
+            "id": device_id,
+            "model_id": model_id,
```

Comment:
you also need to specific the `client_type` (`SDK_LIBRARY` or `SDK_SERVICE`):
https://developers.google.com/assistant/sdk/reference/device-registration/model-and-instance-schemas#device_instance_json

Reply:
From what I see in the docs, it looks like `clientType` is Optional?

Reply:
I've specified it as SDK_SERVICE to stay on the safe side (the assistant library will create its own device with SDK_LIBRARY).